[Flang][OpenMP] Compilation error when type-list in declare reduction directive is a derived type accessed by use association

```
Version of flang : 22.0.0(2f7e218017db69454ea84e1849fcb9a11b1f7fef)/AArch64
```

In the attached program (`masa_udr0009_5.f90`), specifying `derived type name` in `type-list` of `declare reduction directive` by `use association` causes an error during compilation.
After checking `OpenMP 6.0 : 7.6.14 declare_reduction Directive`, the program appears to contain no particular errors.

The following are the test program, Flang, Gfortran and ifx compilation/execution results.

masa_udr0009_5.f90:
```fortran
module mod
  type t
     integer::i=0
  end type t
!$omp declare reduction (+:t:omp_out%i=omp_out%i+omp_in%i) &
!$omp         initializer(omp_priv%i=0)
end module mod

program main
  use mod
  integer::i
  type(t)::x1
  x1%i=0
!$omp parallel do reduction(+:x1)
  do i=1,10
     x1%i=x1%i+1
  end do
!$omp end parallel do
  print *,'pass'
end program main
```

```
$ flang -fopenmp masa_udr0009_5.f90
error: Semantic errors in masa_udr0009_5.f90
./masa_udr0009_5.f90:14:31: error: The type of 'x1' is incompatible with the reduction operator.
  !$omp parallel do reduction(+:x1)
                                ^^
$
```

```
$ export OMP_NUM_THREADS=2; gfortran -fopenmp masa_udr0009_5.f90; ./a.out
 pass
$
```

```
$ export OMP_NUM_THREADS=2; ifx -qopenmp masa_udr0009_5.f90; ./a.out
 pass
$
```

--- a/flang/include/flang/Semantics/symbol.h
+++ b/flang/include/flang/Semantics/symbol.h
@@ -764,6 +764,22 @@ public:
         return true;
       }
     }
+    // For derived and class-derived types, match on the type symbol pointer.
+    if (type.category() == DeclTypeSpec::TypeDerived ||
+        type.category() == DeclTypeSpec::ClassDerived) {
+      const auto &rhs = type.derivedTypeSpec();
+      const auto &rhsSym = rhs.typeSymbol();
+      for (auto t : typeList_) {
+        if (t->category() == DeclTypeSpec::TypeDerived ||
+            t->category() == DeclTypeSpec::ClassDerived) {
+          const auto &lhs = t->derivedTypeSpec();
+          const auto &lhsSym = lhs.typeSymbol();
+          if (&lhsSym == &rhsSym) {
+            return true;
+          }
+        }
+      }
+    }
     return false;
   }
 

--- a/flang/lib/Semantics/check-omp-structure.cpp
+++ b/flang/lib/Semantics/check-omp-structure.cpp
@@ -3691,6 +3691,20 @@ static bool CheckSymbolSupportsType(const Scope &scope,
       return reductionDetails->SupportsType(type);
     }
   }
+  // Look through module scopes in the global scope.
+  // This covers reductions declared in a module and used via USE association
+  const SemanticsContext &semCtx{scope.context()};
+  Scope &global = const_cast<SemanticsContext &>(semCtx).globalScope();
+  for (const Scope &child : global.children()) {
+    if (child.kind() == Scope::Kind::Module) {
+      if (const auto *symbol{child.FindSymbol(name)}) {
+        if (const auto *reductionDetails{
+                symbol->detailsIf<UserReductionDetails>()}) {
+          return reductionDetails->SupportsType(type);
+        }
+      }
+    }
+  }
   return false;
 }
 

--- a/flang/test/Semantics/OpenMP/declare-reduction-derived-module.f90
+++ b/flang/test/Semantics/OpenMP/declare-reduction-derived-module.f90
@@ -1,0 +1,30 @@
+! RUN: %flang_fc1 -fdebug-dump-symbols -fopenmp -fopenmp-version=50 %s | FileCheck %s
+module mod
+  type t
+     integer::i=0
+  end type t
+!$omp declare reduction (+:t:omp_out%i=omp_out%i+omp_in%i) &
+!$omp         initializer(omp_priv%i=0)
+end module mod
+
+!CHECK: Module scope: mod
+!CHECK: op.+, PUBLIC: UserReductionDetails TYPE(t)
+!CHECK: t, PUBLIC: DerivedType components: i
+
+program main
+  use mod
+  integer::i
+  type(t)::x1
+  x1%i=0
+!$omp parallel do reduction(+:x1)
+  do i=1,10
+     x1%i=x1%i+1
+  end do
+!$omp end parallel do
+  print *,'pass'
+end program main
+
+!CHECK: MainProgram scope: MAIN
+!CHECK: op.+: Use from op.+ in mod
+!CHECK: t: Use from t in mod
+!CHECK: x1 size=4 offset=4: ObjectEntity type: TYPE(t)

